File: Algorithms/Medium/PartitionList.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        less, more = ListNode(), ListNode()
        l, m = less, more
        while head:
            if head.val < x:
                l.next = head
                l = l.next
            else:
                m.next = head
                m = m.next

            head = head.next

        l.next = m.next = None
        l.next = more.next
        return less.next

File: Algorithms/Medium/PartitionList.go
package medium

import "github.com/parviz-yu/LeetCode/Algorithms/helpers"

// Time complexity: O(n)
// Space complexity: O(1)
func partition(head *helpers.ListNode, x int) *helpers.ListNode {
	less, more := helpers.ListNode{}, helpers.ListNode{}
	l_ptr, m_ptr := &less, &more

	for head != nil {
		if head.Val < x {
			l_ptr.Next = head
			l_ptr = l_ptr.Next
		} else {
			m_ptr.Next = head
			m_ptr = m_ptr.Next
		}

		head = head.Next
	}

	m_ptr.Next = nil
	l_ptr.Next = more.Next
	return less.Next
}
